doc/man/libdontdie.7
.\" 
.\" Man page for libdontdie
.\"
.\" For license, see the 'LICENSE' file.
.\"
.TH libdontdie 7 2015-03-07 "TCP keepalive support lib" "TCP keepalive support lib"
.SH NAME
libdontdie \- a library that sets the TCP keep-alive flag when
applications call socket(2)
.SH DESCRIPTION
.B libdontdie
is a shared library that can be injected into any
application with the LD_PRELOAD mechanism. It is completely
transparent to the application and therefore works equally well with
closed-source programs and with languages like Java that do not
natively allow changing the TCP keep-alive parameters. 
.SH BACKGROUND
In theory, a TCP connection that is not explicitly closed remains open
forever.  In practice, this does not work out when connection-tracking
firewalls along the way drop their state, and all further traffic, for
a connection that was idle for a certain time.  The result is that
both the TCP client and server believe the connection to be open but
the firewall inhibits all further communication.
.P
This kind of connection failure will be reported to the participating
applications only after they attempt to send further data and all TCP
timeouts and retransmits expire.
.P
In situation where only one of the communication partners ever sends
anything, e.g. the client waiting for HTTP chunked data, it can be
impossible to distinguish between a connection closed by a firewall
and connected-but-idle.
.P
To handle and/or prevent this kind of situation, TCP has a keep-alive
mechanism.  It consists of TCP packets sent periodically over an
otherwise idle connection to "refresh" the associated state along the
whole path.  If not acknowledged the same retry-mechanism is used as
for regular data packets, however independent of whether there is any
data to send. Finally all unanswered keep-alive packets leads to a TCP
error that can be detected by the application. 
.P
By default the first TCP KEEP ALIVE packet is
send after two hours, which is too long for most scenarios.
.P
The Linux kernel offers the possibility to change the TCP keep-alive
timeout globally for the whole system.  However this solution requires
root privileges and does not work when different applications require
different timeouts on the same machine.
.P
In some programming languages an appropriate API call can be used to
enable TCP keep-alive (e.g. setsockopt(2) in C).  Other languages like
Java do not support changing these parameters.
.P
.B libdontdie
can be used for applications that are written in languages that do not
support setting the TCP KEEP ALIVE parameters.  It can also be used to
enable TCP keep-alive for closed-source or commercial applications.
.B libdontdie
can be used on a case-by-case basis, selectively enabling
TCP keep-alive for some application instances without changing any
source code. All parameters are passed to libdontdie in the form of
environment variables. 
.SH TCP KEEP ALIVE PARAMETERS
Three parameters are used to change the TCP keep-alive behaviour. All
parameters are in seconds.
.P
The variable names are taken from the Linux kernel configuration with
a 'DD_' prefix for 'libDontDie'. The corresponding kernel parameters
can be found in '/proc/sys/net/ipv4/'. 
.SS DD_TCP_KEEPALIVE_TIME
The time between the last TCP data packet sent and the first TCP KEEP
ALIVE packet / probe.
.SS DD_TCP_KEEPALIVE_INTVL
The interval between two TCP KEEP ALIVE packets / probes.
.SS DD_TCP_KEEPALIVE_PROBES
The number of keep-alive probes sent before the socket enters an error
state.
.SH LIBDONTDIE PARAMETERS
Two additional parameters change the behaviour of libdontdie itself:
.B libdontdie:
.SS DD_DEBUG
If this is set to '1', each 'socket(2)' call will be logged
to syslog - including all parameters and actions performed by
.B libdontdie.
.SS DD_EVAL_ENVIRONMENT_ONCE
If this is set to '1' or not specified at all, all parameters are only
evaluated once during startup.  If this parameter is set to '0', each
time a socket call is executed, all parameters are evaluated again.
This makes it possible to change parameters at runtime.  When setting
this to '0', there is a bigger per socket call overhead therefore
the performance will decrease.
.P
In languages that do not support setting the TCP KEEP ALIVE
parameters, this is a workaround to enable different setting for
different sockets.
.sp
.RS
.nf
	setenv("DD_TCP_KEEPALIVE_TIME", 60);
	socket(...);
	....
	setenv("DD_TCP_KEEPALIVE_TIME", 180);
	socket(...);
.SH INSTALLATION LOCATION
Depending on the installation method or the distribution, the
installation directory of
.B libdontdie
might differ.
.P
One way to get the installation directory is using the packet manager
to list all files of the packet (like 'dpkg -L <package_name>').
.P
Typically the library is installed in a directory under /usr/lib,
/usr/lib64, or /usr/lib/<triple>.  A typical triple
is 'x86_64-linux-gnu'.  Under Debian it is possible to get
the triple with the command
.sp
.RS
.nf
	dpkg-architecture -qDEB_HOST_GNU_TYPE
.SH USAGE
All parameters are passed in as environment variables.  The
.B libdontdie
itself is preloaded.
.P
The example assumes, that the library is installed
under '/usr/lib/libdontdie.so'.  This might be replaced by the real
installation path.
.P
Example: to run the java program EchoClient with special TCP KEEP
ALIVE setting, use:
.sp
.RS
.nf
DD_DEBUG=1 DD_TCP_KEEPALIVE_TIME=4 DD_TCP_KEEPALIVE_INTVL=5 \\ 
   DD_TCP_KEEPALIVE_PROBES=6 LD_PRELOAD=/usr/lib/libdontdie.so \\ 
   java EchoClient 127.0.0.1 22
.SH "SEE ALSO"
.BR socket(2),
.BR setsockopt(2)
.SH HISTORY
The idea was first implemented in libkeepalive by Fabio Busatto.
Because of some limitations regarding functionality and license, it
was completely rewritten, corrected and extended.
.SH AUTHOR
Written by Andreas Florath (andreas@florath.net)
.SH COPYRIGHT
Copyright \(co 2015 by Andreas Florath (andreas@florath.net).
License MIT.


debian/changelog
libdontdie (1.0-1) unstable; urgency=low

  * Initial release

 -- Andreas Florath <andreas@florath.net>  Wed, 04 Mar 2015 08:08:00 +0200


src/libdontdie.c
#define _GNU_SOURCE

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <dlfcn.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/tcp.h>

#define LOG(msg...)                                                            \
  do {                                                                         \
    if (debug > 0) {                                                           \
      syslog(LOG_USER | LOG_INFO, "libdontdie: " msg);                         \
    }                                                                          \
  } while (0)

int (*socket_call_clib)(int domain, int type, int protocol);

int debug = 0;
int use_keepalive = 1;
int eval_environment_once = 0;
int tcp_keepalive_time = -1;
int tcp_keepalive_intvl = -1;
int tcp_keepalive_probes = -1;

static void eval_environment() {
  if (eval_environment_once) {
    return;
  }

  char const *const str_debug = getenv("DD_DEBUG");
  if (str_debug != NULL && str_debug[0] == '1') {
    debug = 1;
  }

  LOG("Initialization");

  char const *const str_eval_environment_once =
      getenv("DD_EVAL_ENVIRONMENT_ONCE");
  if (str_eval_environment_once != NULL && *str_eval_environment_once == '0') {
    LOG("Evaluate environment for every socket() call");
  } else {
    eval_environment_once = 1;
    LOG("Evaluate environment only once");
  }

  char const *const str_use_keepalive = getenv("DD_USE_KEEPALIVE");
  if (str_use_keepalive != NULL && *str_use_keepalive == '0') {
    use_keepalive = 0;
    LOG("TCP keepalive is switched off");
  } else {
    LOG("TCP keepalive is switched on");
  }

#define EVAL_ENV(ltype, strtype)                                               \
  char const *const str_tcp_keepalive_##ltype =                                \
      getenv("DD_TCP_KEEPALIVE_" strtype);                                     \
  if (str_tcp_keepalive_##ltype != NULL) {                                     \
    tcp_keepalive_##ltype = atoi(str_tcp_keepalive_##ltype);                   \
    LOG("set TCP_KEEPALIVE_" strtype " [%d]", tcp_keepalive_##ltype);         \
  }

  EVAL_ENV(time, "TIME");
  EVAL_ENV(intvl, "INTVL");
  EVAL_ENV(probes, "PROBES");

#undef EVAL_ENV
}

__attribute__((constructor)) void libdontdie_init() {
  eval_environment();

  // Cast it: accoring to the man page this construct must be used:
  *(void **)(&socket_call_clib) = dlsym(RTLD_NEXT, "socket");
  if (socket_call_clib == NULL) {
    LOG("No dynamic symbol with name 'socket' found [%m]");
    abort();
  }
}

static void log_parameter_domain(int const domain) {
  switch (domain) {
  case AF_UNIX:
    LOG("domain [AF_UNIX]");
    break;
  case AF_INET:
    LOG("domain [AF_INET]");
    break;
  case AF_INET6:
    LOG("domain [AF_INET6]");
    break;
  case AF_IPX:
    LOG("domain [AF_IPX]");
    break;
  case AF_NETLINK:
    LOG("domain [AF_NETLINK]");
    break;
  case AF_X25:
    LOG("domain [AF_X25]");
    break;
  case AF_AX25:
    LOG("domain [AF_AX25]");
    break;
  case AF_ATMPVC:
    LOG("domain [AF_ATMPVC]");
    break;
  case AF_APPLETALK:
    LOG("domain [AF_APPLETALK]");
    break;
  case AF_PACKET:
    LOG("domain [AF_PACKET]");
    break;
  }
}

static void log_parameter_type(int const type) {
  if (type & SOCK_STREAM) {
    LOG("type [SOCK_STREAM]");
  }
  if (type & SOCK_DGRAM) {
    LOG("type [SOCK_DGRAM]");
  }
  if (type & SOCK_SEQPACKET) {
    LOG("type [SOCK_SEQPACKET]");
  }
  if (type & SOCK_RAW) {
    LOG("type [SOCK_RAW]");
  }
  if (type & SOCK_RDM) {
    LOG("type [SOCK_RDM]");
  }
  if (type & SOCK_PACKET) {
    LOG("type [SOCK_PACKET]");
  }
  if (type & SOCK_NONBLOCK) {
    LOG("type [NONBLOCK]");
  }
  if (type & SOCK_CLOEXEC) {
    LOG("type [CLOEXEC]");
  }
}

static void log_parameters(int domain, int type, int protocol) {
  log_parameter_domain(domain);
  log_parameter_type(type);
  LOG("protocol [%d]", protocol);
}

int socket(int domain, int type, int protocol) {
  eval_environment();
  if (debug) {
    LOG("socket() called");
    log_parameters(domain, type, protocol);
  }

  int const socket_fd = (*socket_call_clib)(domain, type, protocol);
  if (socket_fd == -1) {
    LOG("Error socket() call [%m]");
    return socket_fd;
  }
  LOG("socket() call returned fd [%d]", socket_fd);

  // Check some parameters
  if (domain != AF_INET && domain != AF_INET6) {
    LOG("Ignoring TCP KEEPALIVE on a non INET4/6 socket");
    return socket_fd;
  }
  if ((type & SOCK_STREAM) == 0) {
    LOG("Ignoring TCP KEEPALIVE on a non stream socket");
    return socket_fd;
  }

  LOG("Parameters check passed");

  if (use_keepalive == 0) {
    LOG("Keepalive switched off by configuration");
    return socket_fd;
  }

  LOG("Setting KEEPALIVE for socket");

  int const ssopt_ka = setsockopt(socket_fd, SOL_SOCKET, SO_KEEPALIVE,
                                  &use_keepalive, sizeof(use_keepalive));
  if (ssopt_ka == -1) {
    LOG("Setting KEEPALIVE returned error [%m]");
    return socket_fd;
  }

#define SET_SOCK_OPT(ltype, strtype, tcp_type)                                 \
  if (tcp_keepalive_##ltype >= 0) {                                            \
    LOG("Seting " strtype " [%d]", tcp_keepalive_##ltype);                     \
    int const ssopt =                                                          \
        setsockopt(socket_fd, SOL_TCP, tcp_type, &tcp_keepalive_##ltype,       \
                   sizeof(tcp_keepalive_##ltype));                             \
    if (ssopt == -1) {                                                         \
      LOG("Setting " strtype " returned error [%m]");                          \
    }                                                                          \
  }

  SET_SOCK_OPT(time, "TIME", TCP_KEEPIDLE);
  SET_SOCK_OPT(intvl, "INTVL", TCP_KEEPINTVL);
  SET_SOCK_OPT(probes, "PROBES", TCP_KEEPCNT);

#undef SET_SOCK_OPT

  LOG("Finished; returning to caller [%d]", socket_fd);

  return socket_fd;
}
